subsys/net/lib/lwm2m/lwm2m_shell.c

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#define LOG_MODULE_NAME net_lwm2m_shell
+#define LOG_LEVEL CONFIG_LWM2M_LOG_LEVEL
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#include <stddef.h>
+#include <stdlib.h>
+
+#include <kernel.h>
+#include <net/lwm2m.h>
+#include <net/lwm2m_path.h>
+#include <shell/shell.h>
+#include <sys/atomic.h>
+
+#include "lwm2m_engine.h"
+#include "lwm2m_rd_client.h"
+
+#define LWM2M_HELP_CMD "LwM2M commands"
+#define LWM2M_HELP_SEND "LwM2M SEND operation"
+
+#define PR(fmt, ...) shell_fprintf(shell, SHELL_NORMAL, fmt, ##__VA_ARGS__)
+
+#define PR_SHELL(shell, fmt, ...) shell_fprintf(shell, SHELL_NORMAL, fmt, ##__VA_ARGS__)
+
+#define PR_ERROR(fmt, ...) shell_fprintf(shell, SHELL_ERROR, fmt, ##__VA_ARGS__)
+
+#define PR_INFO(fmt, ...) shell_fprintf(shell, SHELL_INFO, fmt, ##__VA_ARGS__)
+
+#define PR_WARNING(fmt, ...) shell_fprintf(shell, SHELL_WARNING, fmt, ##__VA_ARGS__)
+
+static struct lwm2m_ctx *cctx;
+static char const **send_paths;
+static size_t path_count;
+
+int lwm2m_shell_init(struct lwm2m_ctx *client_ctx, char const **paths, size_t path_cnt)
+{
+	if (cctx) {
+		return -EALREADY;
+	}
+
+	cctx = client_ctx;
```

Comment:
I just wonder, instead of having the application to set the context, we could add a helper function to retrieve the currently used `lwm2m_ctx` from `lwm2m_rd_client` - the engine only supports a single instance at a time anyway. That way we might not have to initialize/deinitialize the shell module at all.

Reply:
The first part sounds good, I'll do that. I still do think that the app should provide the path literals though.

Reply:
Fixed.